doc/api/async_hooks.md

```diff
@@ -0,0 +1,677 @@
+
+* Returns {AsyncHook} A reference to `asyncHook`.
+
+Disable the callbacks for a given `AsyncHook` instance from the global pool of
+hooks to be executed. Once a hook has been disabled it will not fire again
+until enabled.
+
+For API consistency `disable()` also returns the `AsyncHook` instance.
+
+#### Hook Callbacks
+
+Key events in the lifetime of asynchronous events have been categorized into
+four areas: instantiation, before/after the callback is called, and when the
+instance is destructed. For cases where resources are reused, instantiation and
+destructor calls are emulated.
+
+##### `init(id, type, triggerId, resource)`
+
+* `id` {number} a unique id for the async resource
+* `type` {string} the type of the async resource
+* `triggerId` {number} the unique id of the async resource in whose
+  execution context this async resource was created
+* `resource` {Object} reference to the resource representing the async operation,
+  needs to be released during _destroy_
+
+Called when a class is constructed that has the _possibility_ to trigger an
+asynchronous event. This _does not_ mean the instance must trigger
+`before()`/`after()` before `destroy()` is called. Only that the possibility
+exists.
+
+This behavior can be observed by doing something like opening a resource then
+closing it before the resource can be used. The following snippet demonstrates
+this.
+
+```js
+require('net').createServer().listen(function() { this.close() });
+// OR
+clearTimeout(setTimeout(() => {}, 10));
+```
+
+Every new resource is assigned a unique id.
+
+The `type` is a String that represents the type of resource that caused
+`init()` to fire. Generally it will be the name of the resource's constructor.
+Some examples include `TCP`, `GetAddrInfo` and `HTTPParser`. Users will be able
+to define their own `type` when using the public embedder API.
+
+*Note:* It is possible to have type name collisions. Embedders are recommended
+to use unique prefixes per module to prevent collisions when listening to the
+hooks.
+
+`triggerId` is the `id` of the resource that caused (or "triggered") the
+new resource to initialize and that caused `init()` to fire.
+
+The following is a simple demonstration of this:
+
+```js
+const async_hooks = require('async_hooks');
+
+async_hooks.createHook({
+  init (id, type, triggerId) {
+    const cId = async_hooks.currentId();
+    process._rawDebug(`${type}(${id}): trigger: ${triggerId} scope: ${cId}`);
+  }
+}).enable();
+
+require('net').createServer(c => {}).listen(8080);
+```
+
+Output when hitting the server with `nc localhost 8080`:
+
+```
+TCPWRAP(2): trigger: 1 scope: 1
+TCPWRAP(4): trigger: 2 scope: 0
+```
+
+The second `TCPWRAP` is the new connection from the client. When a new
+connection is made the `TCPWrap` instance is immediately constructed. This
+happens outside of any JavaScript stack (side note: a `currentId()` of `0`
+means it's being executed in "the void", with no JavaScript stack above it).
```

Comment:
This should be clarified that "the void" means coming from an async operation in a C++ addon (IIRC).